### PR TITLE
Add Checking of Bid Public Keys

### PR DESCRIPTION
**Pull request by PatStiles**
WIP for  #82 . 

- [x] Extends Relay Client to tracks BlsPublicKeys
- [x] Checks that a Relays BlsPublicKey matches SignedBuilder BlsPublicKey
- [x] Parses BlsPublicKey from Relay Mux config. This should probably involve adding a FromStr implementation to https://github.com/ralexstokes/ethereum-consensus/blob/main/src/crypto.rs on top of the other trait implementations for PublicKey on the next release.

### CODE REVIEW

File: mev-boost-rs/src/relay_mux.rs [hunk 1]
```diff
@@ -112,11 +112,16 @@ impl BlindedBlockProvider for RelayMux {
...
             .filter_map(|(relay_index, response)| match response {
                 Ok(mut bid) => {
-                    if let Err(err) = validate_bid(&mut bid, &self.context) {
-                        tracing::warn!("invalid signed builder bid: {err} for bid: {bid:?}");
+                    if &self.relays[relay_index].public_key != bid.public_key() {
+                        tracing::warn!("invalid public key for bid: {bid:?}");
                         None
                     } else {
-                        Some((bid, relay_index))
+                        if let Err(err) = validate_bid(&mut bid, &self.context) {
```
Comment (ralexstokes) [inline mev-boost-rs/src/relay_mux.rs:119]: better to encapsulate this inside `validate_bid`

just pass the `&self.relays[relay_index]` as another parameter to `validate_bid` and check there

File: mev-boost-rs/src/relay_mux.rs [hunk 2]
```diff
@@ -206,6 +211,9 @@ fn bid_key_from(signed_block: &SignedBlindedBeaconBlock, public_key: &BlsPublicK
 mod tests {
     use super::*;
 
+    #[test]
+    fn test_bid_verification() {}
```
Comment (ralexstokes) [inline mev-boost-rs/src/relay_mux.rs:215]: remove this unless you want to add a test -- which would be great!

File: mev-boost-rs/src/relay_mux.rs [hunk 3]
```diff
@@ -129,7 +134,7 @@ impl BlindedBlockProvider for RelayMux {
         let best_indices = select_best_bids(bids.iter().map(|(bid, i)| (bid.value(), *i)));
 
         if best_indices.is_empty() {
-            return Err(Error::NoBids)
+            return Err(Error::NoBids);
```
Comment (ralexstokes) [inline mev-boost-rs/src/relay_mux.rs:137]: is from the nightly cargo fmt?

try `just fmt` from the repo root dir to see if it stays

File: mev-boost-rs/src/service.rs [hunk 4]
```diff
@@ -107,7 +123,7 @@ impl Future for ServiceHandle {
         let this = self.project();
         let relay_mux = this.relay_mux.poll(cx);
         if relay_mux.is_ready() {
-            return relay_mux
+            return relay_mux;
```
Comment (ralexstokes) [inline mev-boost-rs/src/service.rs:126]: same thing on formatting change here

File: mev-boost-rs/src/service.rs [hunk 5]
```diff
@@ -45,21 +47,33 @@ fn parse_url(input: &str) -> Option<Url> {
...
+    } else {
+        Some(BlsPublicKey::default())
+    }
+}
+
 pub struct Service {
     host: Ipv4Addr,
     port: u16,
-    relays: Vec<Url>,
+    relays: Vec<(Url, BlsPublicKey)>,
```
Comment (ralexstokes) [inline mev-boost-rs/src/service.rs:62]: at this point I'd move to encapsulate these details behind a `Relay` type

I'd do this:

1) add `relay.rs` along with a `mod relay;` in the `lib.rs`
2) define something like:
```rust
pub struct Relay {
    api: BlindedBlockProviderClient,
    public_key: BlsPublicKey,
}
```
3. parse the config `Strings` as URLs, make a client for each URL, parse the public key from the URL's `hostname` and then use some combo of `hex::decode` and `BlsPublicKey::try_from(_: &[u8])` to get a key; provide the client and the key as inputs to a `Relay::new`

File: mev-boost-rs/src/service.rs [hunk 6]
```diff
@@ -16,6 +16,7 @@ pub struct Config {
     pub host: Ipv4Addr,
     pub port: u16,
     pub relays: Vec<String>,
+    pub public_keys: Vec<String>,
```
Comment (ralexstokes) [inline mev-boost-rs/src/service.rs:19]: the format has the public key _in_ the URL which is already provided in the `relays` entry to this config; so we can remove this -- see the comment below about a way to grab the keys from the URLs themselves while making `Relay` instances

**Review by ralexstokes — CHANGES_REQUESTED** [review]
nice start! I left some comments to iterate the code so we end up with better encapsulation of all the details of handling relay data

**Comment by PatStiles** [comment]
resolves #82

**Review by ralexstokes — CHANGES_REQUESTED** [review]
nice work! left some comments/suggestions for how to move forward

**Comment by ralexstokes** [comment]
fixed in #96, closing as stale, thank you for the effort here!

**Pull request closed without merging**